README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# pmxpartab

<!-- badges: start -->
<!-- [![experimental](https://badges.github.io/stability-badges/dist/experimental.svg)](http://github.com/badges/stability-badges) -->
<!-- [![R-CMD-check](https://github.com/benjaminrich/pmxpartab/workflows/R-CMD-check/badge.svg)](https://github.com/benjaminrich/pmxpartab/actions) -->

[![CRAN\_Release\_Badge](https://www.r-pkg.org/badges/version-ago/pmxpartab)](https://CRAN.R-project.org/package=pmxpartab)
<!-- [![CRAN\_Download\_Badge](https://cranlogs.r-pkg.org/badges/pmxpartab)](https://CRAN.R-project.org/package=pmxpartab) -->

<!-- badges: end -->

This R package produces nice looking parameter tables for pharmacometric
modeling results with ease. It is completely agnostic to the modeling
software that was used.

## Installation

This package is not yet on CRAN. To install the latest development
version directly from GitHub:

``` r
require(remotes)
remotes::install_github("benjaminrich/pmxpartab")
```

## Usage

The creation of the parameter table proceeds in 2 steps:

1.  Generate an intermediate `data.frame` from model outputs and
    metadata.
2.  Generate a formatted HTML table from that intermediate `data.frame`.

Both the model outputs and metadata are provided as R lists (it is a
separate problem to extract the outputs from the modeling software into
the required list format).

Here is an example (in this example, YAML is used to give a clear and
concise representation of outputs and metadata, but this is not
required).

``` r
library(yaml)

outputs <- yaml.load("
est:
  CL:    0.482334
  VC:    0.0592686
  CL_WT: 0.750000
  VC_WT: 1.00000
  nCL:   0.315414
  nVC:   0.536025
  ERRP:  0.0508497
se:
  CL:    0.0138646
  VC:    0.0055512
  nCL:   0.0188891
  nVC:   0.0900352
  ERRP:  0.0018285
fixed:
  CL:    no
  VC:    no
  CL_WT: yes
  VC_WT: yes
  nCL:   no
  nVC:   no
  ERRP:  no
shrinkage:
  nCL:  9.54556
  nVC:  47.8771
")

meta <- yaml.load("
parameters:
- name:  CL
  label: 'Clearance'
  units: 'L/h'
  type:  Structural

- name:  VC
  label: 'Volume'
  units: 'L'
  type:  Structural
  trans: 'exp'
  
- name:  CL_WT
  label: 'Weight on Clearance'
  type:  CovariateEffect

- name:  VC_WT
  label: 'Weight on Volume'
  type:  CovariateEffect
  
- name:  nCL
  label: 'On Clearance'
  type:  IIV
  trans: 'SD (CV%)'
  
- name:  nVC
  label: 'On Volume'
  type:  IIV
  trans: 'SD (CV%)'
  
- name:  ERRP
  label: 'Proportional Error'
  units: '%'
  type:  RUV
  trans: '%'
")

parframe <- pmxparframe(outputs, meta)
parframe
#>    name               label units            type    trans fixed      est
#> 1    CL           Clearance   L/h      Structural     <NA> FALSE 0.482334
#> 2    VC              Volume     L      Structural      exp FALSE 1.061060
#> 3 CL_WT Weight on Clearance  <NA> CovariateEffect     <NA>  TRUE 0.750000
#> 4 VC_WT    Weight on Volume  <NA> CovariateEffect     <NA>  TRUE 1.000000
#> 5   nCL        On Clearance  <NA>             IIV SD (CV%) FALSE 0.315414
#> 6   nVC           On Volume  <NA>             IIV SD (CV%) FALSE 0.536025
#> 7  ERRP  Proportional Error     %             RUV        % FALSE 5.084970
#>            se       rse     lci95     uci95         pval shrinkage
#> 1 0.013864600  2.874481 0.4551594 0.5095086 0.000000e+00        NA
#> 2 0.005890157  0.555120 1.0495781 1.0726679 0.000000e+00        NA
#> 3          NA        NA        NA        NA           NA        NA
#> 4          NA        NA        NA        NA           NA        NA
#> 5 0.018889100  5.988669 0.2783914 0.3524366 0.000000e+00   9.54556
#> 6 0.090035200 16.796829 0.3595560 0.7124940 2.624601e-09  47.87710
#> 7 0.182850000  3.595891 4.7265840 5.4433560 0.000000e+00        NA
```

``` r
pmxpartab(parframe)
```

Which produces:

![Example result: parameter
table](tools/readme/pmxpartab-example-output.png)

For more information, read the
[vignette](https://benjaminrich.github.io/pmxpartab/vignettes/pmxpartab-vignette.html).


NEWS.md
# `pmxpartab` v0.5.0

* This is the initial release of `pmxpartab` on CRAN.



R/functions.R
#' Get parameters in a data.frame
#'
#' @export
parframe <- function(outputs, meta=outputs$meta) {
    param <- meta$parameters
    z <- data.table::rbindlist(param, fill=T)

    outputs$all       <- with(outputs, c(th, om, sg))
    outputs$se$all    <- with(outputs$se, c(th, om, sg))
    outputs$fixed$all <- with(outputs$fixed, c(th, om, sg))

    z$fixed     <- as.logical(NA)
    z$est       <- as.numeric(NA)
    z$se        <- as.numeric(NA)
    z$rse       <- as.numeric(NA)
    z$lci95     <- as.numeric(NA)
    z$uci95     <- as.numeric(NA)
    z$pval      <- as.numeric(NA)
    z$shrinkage <- as.numeric(NA)

    have.bootstrap <- !is.null(outputs$bootstrap)
    if (have.bootstrap) {
        z$boot.median <- as.numeric(NA)
        z$boot.lci    <- as.numeric(NA)
        z$boot.uci    <- as.numeric(NA)
    }

    for (i in 1:nrow(z)) {

        `%||%` <- function(x, y) { if (is.null(x) || is.na(x)) y else x }
        name <- z$name[i] %||% NA
        trans <- z$trans[i] %||% NA

        if (have.bootstrap) {
            boot.median <- NA
            boot.lci <- NA
            boot.uci <- NA
        }

        # Check parameter type
        if (name %in% names(outputs$all)) {
            est <- outputs$all[[name]]
            se <- outputs$se$all[[name]]
            fixed <- outputs$fixed$all[[name]]
            if (have.bootstrap && name %in% names(outputs$bootstrap$median)) {
                boot.median <- outputs$bootstrap$median[[name]]
                boot.lci <- outputs$bootstrap$ci95[[name]][1]
                boot.uci <- outputs$bootstrap$ci95[[name]][2]
            }
        } else if ((re <- regexec("^om(?:_cor)?\\((\\w+),(\\w+)\\)$", name, perl=T))[[1]][1] > 0) {
            m1 <- regmatches(name, re)[[1]][2]
            m2 <- regmatches(name, re)[[1]][3]
            om_names <- names(outputs$om)
            a1 <- match(m1, om_names)
            a2 <- match(m2, om_names)
            if (!is.na(a1) && !is.na(a2)) {
                est <- outputs$om_cor[a1,a2]
                se <- outputs$se$om_cor[a1,a2]
                fixed <- outputs$fixed$om_cor[a1,a2]
                if (fixed) {
                    se <- NA
                }
                if (have.bootstrap) {
                    boot.re <- paste0("OMEGA\\D", max(a1, a2), "\\D", min(a1, a2), "($|\\D)")
                    boot.name <- grep(boot.re, names(nmout$bootstrap$median), ignore.case=T, value=T)
                    if (length(boot.name) == 1) {
                        boot.median <- outputs$bootstrap$median[[boot.name]]
                        boot.lci <- outputs$bootstrap$ci95[[boot.name]][1]
                        boot.uci <- outputs$bootstrap$ci95[[boot.name]][2]
                    }
                }
            } else {
                # Not found
                next
            }
        } else if ((re <- regexec("^om_cov\\((\\w+),(\\w+)\\)$", name, perl=T))[[1]][1] > 0) {
            a1 <- regmatches(name, re)[[1]][2]
            a2 <- regmatches(name, re)[[1]][3]
            if (a1 %in% dimnames(outputs$om_cov)[[1]] && a2 %in% dimnames(outputs$om_cov)[[2]]) {
                est <- outputs$om_cov[a1,a2]
                se <- outputs$se$om_cov[a1,a2]
                fixed <- outputs$fixed$om_cov[a1,a2]
                if (fixed) {
                    se <- NA
                }
                if (have.bootstrap) {
                    boot.re <- paste0("OMEGA\\D", max(a1, a2), "\\D", min(a1, a2), "($|\\D)")
                    boot.name <- grep(boot.re, names(nmout$bootstrap$median), ignore.case=T, value=T)
                    if (length(boot.name) == 1) {
                        boot.median <- outputs$bootstrap$median[[boot.name]]
                        boot.lci <- outputs$bootstrap$ci95[[boot.name]][1]
                        boot.uci <- outputs$bootstrap$ci95[[boot.name]][2]
                    }
                }
            } else {
                # Not found
                next
            }
        } else {
            # Not found
            next
        }

        if (fixed || is.null(se) || is.na(se)) {
            se <- NA
            rse <- NA
            ci95 <- c(NA, NA)
            pval <- NA
        } else {
            rse <- 100*se/abs(est)
            ci95 <- est + c(-1,1)*1.96*se
            pval <- 2*(1 - pnorm(abs(est/se)))
        }

        # Check transformation
        if (!is.na(trans) && trans == "%") {
            est <- 100*est
            if (!fixed) {
                se  <- 100*se
                ci95  <- 100*ci95
            }
            if (have.bootstrap) {
                boot.median <- 100*boot.median
                boot.lci <- 100*boot.lci
                boot.uci <- 100*boot.uci
            }
        } else if (!is.na(trans) && trans == "exp") {
            est <- exp(est)
            if (!fixed) {
                rse <- 100*se
                se  <- (rse/100)*est
                ci95  <- exp(ci95)
            }
            if (have.bootstrap) {
                boot.median <- exp(boot.median)
                boot.lci <- exp(boot.lci)
                boot.uci <- exp(boot.uci)
            }
        } else if (!is.na(trans) && trans == "ilogit") {
            ilogit <- function(x) { 1 / (1 + exp(-x)) }
            est <- ilogit(est)
            if (!fixed) {
                rse <- 100*se*(1 - est)
                se  <- (rse/100)*est
                ci95  <- ilogit(ci95)
            }
            if (have.bootstrap) {
                boot.median <- ilogit(boot.median)
                boot.lci <- ilogit(boot.lci)
                boot.uci <- ilogit(boot.uci)
            }
        } else if (!is.na(trans) && trans == "CV%") {
            g <- function(x) { 100*sqrt(exp(x^2) - 1) }
            dg <- function(x) { 100*0.5*(1/sqrt(exp(x^2) - 1))*exp(x^2)*2*x }
            x <- est
            est <- g(x)
            if (!fixed) {
                se  <- se*dg(x)
                rse <- 100*se/abs(est)
                ci95  <- g(ci95)
            }
            if (have.bootstrap) {
                boot.median <- g(boot.median)
                boot.lci <- g(boot.lci)
                boot.uci <- g(boot.uci)
            }
        }

        z$fixed[i] <- fixed
        z$est[i]   <- est
        z$se[i]    <- se
        z$rse[i]   <- rse
        z$lci95[i] <- ci95[1]
        z$uci95[i] <- ci95[2]
        z$pval[i]  <- pval
        if (have.bootstrap) {
            z$boot.median[i] <- boot.median
            z$boot.lci[i] <- boot.lci
            z$boot.uci[i] <- boot.uci
        }

        if (z$name[i] %in% names(outputs$shrinkage)) {
            z$shrinkage[i] <- outputs$shrinkage[z$name[i]]
        }
    }
    z <- subset(z, !is.na(est))
    as.data.frame(z)
}

# Internal function to help format numbers
p <- function(x, digits=3, flag="", round.integers=FALSE){
    if (!is.numeric(x)) {
        return(x)
    }
    prefix <- ifelse(flag=="+" & x > 0, "+", "")
    paste0(prefix, table1::signif_pad(x, digits=digits, round.integers=round.integers))
}

parameter.estimate.table.section <- function(label, ncolumns) {
    paste0(c('<tr>',
        paste0(sprintf('<td class="paramsectionheading">%s</td>', c(label, rep("", ncolumns-1))), collapse='\n'),
        '</tr>'), collapse='\n')
}

parameter.estimate.table.row <- function(
    name,
    label          = NULL,
    units          = NULL,
    type           = c("Structural", "CovariateEffect", "IIV", "IOV", "RUV", "Unspecified"),
    trans          = c("identity", "%", "exp", "ilogit", "CV%", "SD (CV%)"),
    expression     = NULL,
    relatedTo      = NULL,
    superscript    = NULL,
    fixed          = NULL,
    est            = NULL,
    se             = NULL,
    rse            = NULL,
    lci95          = NULL,
    uci95          = NULL,
    boot.median    = NULL,
    boot.lci       = NULL,
    boot.uci       = NULL,
    shrinkage      = NULL,
    na             = "n/a",
    digits         = 3,
    indent         = TRUE,
    have.bootstrap = !is.null(boot.median),
    ...) {

    # Check for superscript
    if (is.null(superscript) || is.na(superscript)) {
        superscript <- ""
    } else {
        superscript <- paste0("<sup>", superscript, "</sup>")
    }

    # Check for label
    if (is.null(label) || is.na(label)) {
        label <- name
    }

    # Check for units
    if (!is.null(units) && !is.na(units)) {
        label <- sprintf("%s (%s)", label, units)
    }

    if (!is.null(trans) && !is.na(trans) && trans == "SD (CV%)") {
        g <- function(x) { 100*sqrt(exp(x^2) - 1) }
        x <- est
        est <- sprintf("%s (%s%%)", p(x, digits), p(g(x), digits))
    } else {
        est <- p(est, digits)
    }

    est <- paste0(est, superscript)

    if (fixed) {
        est <- sprintf('%s Fixed', est)
    }

    if (is.na(se)) {
        se <- na
        rse <- na
        ci95 <- na
    } else {
        rse <- p(rse, digits)
        ci95 <- sprintf('%s &ndash; %s', p(lci95, digits), p(uci95, digits))
    }

    if (have.bootstrap) {
        if (is.na(boot.median)) {
            boot.median <- na
            boot.ci95 <- na
        } else {
            boot.median <- p(boot.median, digits)
            boot.ci95 <- sprintf('%s &ndash; %s', p(boot.lci, digits), p(boot.uci, digits))
        }
    } else {
        boot.ci95 <- NULL
    }

    if (!is.null(shrinkage)) {
        if (is.na(shrinkage)) {
            shrinkage <- ""
        } else {
            shrinkage <- sprintf("%s%%", p(shrinkage, digits))
        }
    }
    all <- c(est=est, rse=rse, ci95=ci95, boot.median=boot.median, boot.ci95=boot.ci95, shrinkage=shrinkage)
    paste0(c('<tr>',
        sprintf('<td class="%s">%s</td>', ifelse(isTRUE(indent), "paramlabelindent", "paramlabelnoindent"), label),
        paste0(sprintf('<td>%s</td>', all[names(columns)]), collapse='\n'),
        '</tr>'), collapse='\n')
}

#' Generate a parameter estimates table in HTML
#' 
#' @examples
#' outputs <- list(
#'     th = list(CL = 0.482334, VC = 0.0592686),
#'     om = list(nCL = 0.315414, nVC = 0.536025),
#'     sg = list(ERRP = 0.0508497),
#'     se = list(
#'         th = list(CL = 0.0138646, VC = 0.00555121),
#'         om = list(nCL = 0.0188891, nVC = 0.0900352),
#'         sg = list(ERRP = 0.00182851)),
#'     fixed = list(
#'         th = list(CL = FALSE, VC = FALSE),
#'         om = list(nCL = FALSE, nVC = FALSE),
#'         sg = list(ERRP = FALSE)),
#'     shrinkage = list(nCL = 9.54556, nVC = 47.8771))
#' 
#' meta <- list(
#'     parameters = list(
#'         list(name="CL",   label="Clearance",    units="L/h", type="Structural"),
#'         list(name="VC",   label="Volume",       units="L",   trans="exp", type="Structural"),
#'         list(name="nCL",  label="On Clearance",              trans="SD (CV%)", type="IIV"),
#'         list(name="nVC",  label="On Volume",                 type="IIV"),
#'         list(name="ERRP", label="Proportional Error",        units="%", trans="%", type="RUV")))
#' 
#' parframe(outputs, meta)
#' 
#' pmxpartab(parframe(outputs, meta),
#'     columns=c(est="Estimate", rse="RSE%", ci95="95%CI", shrinkage="Shrinkage"))
#' @export
pmxpartab <- function(
    parframe,

    columns=c(est="Estimate", rse="RSE%", ci95="95%CI", shrinkage="Shrinkage"),

    sections = TRUE,
    section.labels = c(
        Structural      = "Typical Values",
        CovariateEffect = "Covariate Effects",
        RUV             = "Residual Error",
        IIV             = "Between Subject Variability",
        IOV             = "Inter-Occasion Variability"),

    show.fixed.to.zero=F,
    na="n/a",
    digits=3) {

    if (isFALSE(show.fixed.to.zero)) {
        parframe <- subset(parframe, !(fixed & est==0))
    }

    ncolumns <- length(columns) + 1

    thead <- paste0('<tr>\n<th rowspan="2">Parameter</th>\n',
        paste0(paste0('<th rowspan="2">', columns, '</th>'), collapse="\n"), '\n</tr>\n')


    tbody <- ""
    for (i in 1:nrow(parframe)) {
        if (isTRUE(sections)) {
            newsection <- (!is.null(parframe$type) && !is.na(parframe$type[i]) && (i == 1 || parframe$type[i] != parframe$type[i-1]))
            if (newsection) {
                type <- parframe$type[i]
                if (type %in% names(meta$labels)) {
                    label <- meta$labels[[type]]
                } else if (type %in% names(section.labels)) {
                    label <- section.labels[[type]]
                } else {
                    label <- type
                }

                tbody <- paste0(tbody, parameter.estimate.table.section(label, ncolumns=ncolumns), '\n')
            }
        }
        args <- c(parframe[i,], list(na=na, digits=digits, indent=sections))
        tbody <- paste0(tbody, do.call(parameter.estimate.table.row, args), '\n')
    }

    table <- paste0('<table>\n<thead>\n', thead, '\n</thead>\n<tbody>\n', tbody, '\n</tbody>\n</table>\n')
    structure(table, class=c("pmxpartab", "html", "character"), html=TRUE)
}


#' Print \code{pmxpartab} object.
#' @param x An object returned by \code{\link{pmxpartab}}.
#' @param ... Further arguments passed on to other \code{print} methods.
#' @return Returns \code{x} invisibly.
#' @details In an interactive context, the rendered table will be displayed in
#' a web browser. Otherwise, the HTML code will be printed as text.
#' @export
print.pmxpartab <- function(x, ...) {
    if (interactive()) {
        z <- htmltools::HTML(x)
        default.style <- htmltools::htmlDependency("pmxpartab", "1.0",
            src=system.file(package="pmxpartab", "pmxpartab_defaults_1.0"),
            stylesheet="pmxpartab_defaults.css")
        z <- htmltools::div(class="Rpmxpartab", default.style, z)
        z <- htmltools::browsable(z)
        print(z, ...) # Calls htmltools:::print.html(z, ...)
    } else {
        cat(x)
    }
    invisible(x)
}


#' Method for printing in a \code{knitr} context.
#' @param x An object returned by \code{\link{pmxpartab}}.
#' @param ... Further arguments passed on to \code{knitr::knit_print}.
#' @importFrom knitr knit_print
#' @export
knit_print.pmxpartab <- function(x, ...) {
    knit_to_html <-
        !is.null(knitr::opts_knit$get("rmarkdown.pandoc.to")) &&
        grepl("^html", knitr::opts_knit$get("rmarkdown.pandoc.to"))

    if (knit_to_html) {
        z <- htmltools::HTML(x)
        default.style <- htmltools::htmlDependency("pmxpartab", "1.0",
            src=system.file(package="pmxpartab", "pmxpartab_defaults_1.0"),
            stylesheet="pmxpartab_defaults.css")
        z <- htmltools::div(class="Rpmxpartab", default.style, z)
        knitr::knit_print(z, ...)
    } else {
        knitr::knit_print(as.character(x), ...)
    }
}

#' Parse a parameter description string
#'
#' @examples
#' # Example 1: all elements present
#' x <- "CL, label='Clearance', units='L/h', trans=exp, type='Structural'"
#' parse_parameter_description(x)
#' 
#' # Example 2: Some elements missing (trans), will take default value (NULL)
#' x <- "CL, label='Clearance', units='L/h', type='Structural'"
#' parse_parameter_description(x)
#' 
#' # Example 3: Only the name is given
#' x <- "CL"
#' parse_parameter_description(x)
#' 
#' # Example 4: positional arguments
#' x <- "CL, 'Clearance', 'L/h', type='Structural'"
#' parse_parameter_description(x)
#' @export
parse_parameter_description <- function(string) {
    # Returns a structured object representing a description of a parameter
    # (in this example just a list with some attributes; only the name is mandatory)
    parameter_description <- function(name, label=NULL, units=NULL, trans=NULL, type=NULL) {
        list(name=name, label=label, units=units, trans=trans, type=type)
    }

    x <- str2lang(paste0("parameter_description(", string, ")"))
    x[[2]] <- as.character(x[[2]]) # Interpret the first element (name) as a string even if not quoted
    eval(x)
}

